### PR TITLE
Stamp np when out from auction

### PR DESCRIPTION
**Pull request by grandizzy**
- stamp loan t0Np when removed from auction (after repay or take). already stamped when removed from auction after pledge more collateral. Introduced RepayDebtLocalVars struct to hold local vars and to avoid stack too deep error.
- rework mergeOrRemoveCollateral test
- fix tests

### CODE REVIEW

**Review by mattcushman — APPROVED** [review]
lgtm

File: src/base/Pool.sol [hunk 1]
```diff
@@ -51,6 +51,17 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
     uint256 internal poolInitializations;
     mapping(address => mapping(address => mapping(uint256 => uint256))) private _lpTokenAllowances; // owner address -> new owner address -> deposit index -> allowed amount
 
+    struct RepayDebtLocalVars {
+        uint256 borrowerDebt;          // borrower's accrued debt
+        bool    inAuction;             // true if loan still in auction after repay, false otherwise
+        uint256 newLup;                // LUP after auction is taken
+        bool    pull;                  // true if pull action
+        bool    repay;                 // true if repay action
+        bool    restampT0Np;           // true if loan's t0 neutral price should be restamped (when exiting auction)
```
Comment (ith-harvey) [inline src/base/Pool.sol]: suggest renaming this to `stampT0Np` or `updateT0NP`

**Review by ith-harvey — APPROVED** [review]
LGTM

File: src/base/Pool.sol [hunk 2]
```diff
@@ -484,65 +495,66 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
+        vars.pull         = collateralAmountToPull_ != 0;
+        vars.borrowerDebt = Maths.wmul(borrower.t0Debt, poolState.inflator);
 
-        uint256 borrowerDebt = Maths.wmul(borrower.t0Debt, poolState.inflator);
-        // loan can only be in auction when repaying debt
-        // if loan in auction and pull collateral attempted then borrower collateralization check should revert
-        bool inAuction;
-
-        if (repay) {
+        if (vars.repay) {
```
Comment (ith-harvey) [inline src/base/Pool.sol:503]: Wanted to confirm that `restampt0NP` is initialized as false so it should be false if an action is just a `repay`

**Review by ith-harvey — COMMENTED** [review]
Should someone's t0NP be stamped when they are just pledging collateral? -> https://github.com/ajna-finance/contracts/pull/463/files#diff-b8a3c7e9053439915e5f4dc0ecc997ebfa0e8197de2ca64c1a83531c24333a7aR408

I thought this was not intended logic.

File: src/base/Pool.sol (hunk 2 above)
Comment (grandizzy) [reply src/base/Pool.sol:503]: Correct, default for bool is false

**Comment by grandizzy** [comment]
> Should someone's t0NP be stamped when they are just pledging collateral? -> https://github.com/ajna-finance/contracts/pull/463/files#diff-b8a3c7e9053439915e5f4dc0ecc997ebfa0e8197de2ca64c1a83531c24333a7aR408
> 
> I thought this was not intended logic.

Correct, it is always stamped on pledge / borrow

File: src/base/Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/base/Pool.sol]: done in https://github.com/ajna-finance/contracts/pull/463/commits/e3923a8099c92aaef7bf0da070d08ebe50d1809b

**Comment by ith-harvey** [comment]
t0NP is "stamped" on a loan on:
- `repayDebt` is called
- `take` when loan is removed from auction
- `bucketTake` when loan is removed from auction
- `drawDebt` is called for both pledging and drawing debt actions

**Pull request merged**